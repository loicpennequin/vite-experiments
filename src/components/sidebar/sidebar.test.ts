import nock from 'nock';
import {
  httpMock,
  renderWithPlugins,
  setDeviceWidth
} from '../../__test__/utils';
import Sidebar from './sidebar.vue';
import { waitFor, fireEvent } from '@testing-library/vue';

type PokemonEntry = {
  name: String;
};

type RenderWithLoadedListOptions = {
  results?: PokemonEntry[];
  isDesktop?: boolean;
};

async function renderWithLoadedList({
  results = [{ name: 'foo' }],
  isDesktop = false
}: RenderWithLoadedListOptions = {}) {
  if (isDesktop) {
    setDeviceWidth(1920);
  }

  const scope = httpMock({
    url: '/pokemon?limit=50&offset=0',
    response: {
      count: 50,
      results
    }
  });

  expect(Sidebar).toBeTruthy();

  const wrapper = renderWithPlugins(Sidebar);
  await waitFor(() => expect(scope.isDone()).toBeTruthy());

  return wrapper;
}

describe('Sidebar Component', () => {
  test('should mount component', async () => {
    const { html } = await renderWithLoadedList({ isDesktop: true });

    expect(html()).toMatchSnapshot();
  });

  test('should display a link to detail page for each pokemon', async () => {
    const { getByText } = await renderWithLoadedList({ isDesktop: true });

    const link = getByText('foo');

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', '/foo');
  });

  test('should hide the list on mobile', async () => {
    const { queryByText } = await renderWithLoadedList({ isDesktop: false });

    expect(queryByText('foo')).not.toBeInTheDocument();
  });

  test('should display list toggle on mobile', async () => {
    const { getByTitle, queryByText } = await renderWithLoadedList({
      isDesktop: false
    });

    const toggle = getByTitle('Show list');
    expect(toggle).toBeInTheDocument();
  });

  test('should display/hide list when clicking on the toggle', async () => {
    const { getByTitle, queryByText } = await renderWithLoadedList({
      isDesktop: false
    });

    const toggle = getByTitle('Show list');

    await fireEvent.click(toggle);
    const link = queryByText('foo');

    expect(link).toBeInTheDocument();
    expect(toggle).toHaveAttribute('title', 'Hide list');

    await fireEvent.click(toggle);

    expect(link).not.toBeInTheDocument();
    expect(toggle).toHaveAttribute('title', 'Show list');
  });
});
