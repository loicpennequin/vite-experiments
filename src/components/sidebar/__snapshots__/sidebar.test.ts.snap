// Vitest Snapshot v1

exports[`Sidebar Component > should mount component 1`] = `
<nav p-y="3" transition-all="" duration-300="" min-w="3rem" class="expanded" data-v-9a5c856e=""><button title="Show list" md="hidden" data-v-9a5c856e=""> List </button>
  <transition-stub data-v-9a5c856e="">
    <div data-v-9a5c856e="">
      <ul data-v-9a5c856e="">
        <li data-v-9a5c856e=""><a href="/foo" class="" no-underline="" color="inherit" capitalize="" space-x="1" p="3" block="" data-v-9a5c856e=""><span data-v-9a5c856e="">1 -</span> foo</a></li>
      </ul>
      <div class="infinite-scroll-trigger" data-v-9a123074=""></div>
    </div>
  </transition-stub>
</nav>
`;
